bool req_0
bool req_1
bool req_2